### ISSUE TITLE
Add support for polymorphic type handling in Config Beans

### ISSUE THREAD

**Issue by zbiljic** [lightbend/config#507]
I have a use-case which would be solved by having some type of polymorphic type handling.
This means something like [JacksonPolymorphicDeserialization](https://github.com/FasterXML/jackson-docs/wiki/JacksonPolymorphicDeserialization).

This issue is mostly a question whether you would find this useful, and would accept a pull-request for something that solves this.

Note that I have already done this on my forked repo, not as a proof of concept, but ready for a pull-request (https://github.com/zbiljic/config/tree/feature/polymorphic-beans).

All of this is done with minimal changes, only three annotation classes (`ConfigTypeInfo`, `ConfigTypeName`, `ConfigSubTypes`) that closely resemble the ones from Jackson; with minor updates to `ConfigBeanImpl` class. There are tests for all kinds of situations. There are examples for Java and Scala version.

One addition I made that Jackson does not have is ability to load subtypes from `META-INF/services` directory, which should allow defining them in some other place, which would allow some kind of plugin-based system.

If all of this sounds good, I could create a pull-request, and make any changes you suggest/require.

**Comment by zbiljic** [comment]
Excuse me for asking again, but has anyone had the chance to look at this proposal?
I know it's not a minor change (although it is backwards compatible one), but it would mean to me to know if something like this might be added at one point.

**Comment by havocp** [comment]
Hi! Thanks for this. I would love some help from someone who's used both Jackson and Typesafe Config to look at this - if you have code already, maybe make it a PR just so we can see the diff, even if we don't end up merging it as-is?

I feel like it would be quite a bit of work to adequately give you feedback, so that's why I haven't had time to do so yet. 

I'm open to the idea in principle.

**Comment by zbiljic** [comment]
Yes, sorry, thanks for the response.

Even though I implemented everything I could think of (at the time) regarding this, I do not expect to be merged as-is. I added support for a lot of different things, and even though it might be useful (and backward compatible), I feel it is too big of a change to be introduced all at once (at least that's how I would feel if this was my library).

But I would like to at least this be a start of some kind of a RFC for polymorphic deserialization.
